hfile/reader.go

```diff
@@ -47,22 +51,56 @@ func NewReader(file *os.File) (Reader, error) {
 func (hfile *Reader) String() string {
 	return "hfile"
 }
-func getDataBlock(key []byte, blocks *[]DataBlock) (*DataBlock, error) {
-	// TODO(dan): Binary search instead.
-	for i := len(*blocks) - 1; i >= 0; i-- {
-		block := (*blocks)[i]
-		if cmp := bytes.Compare(key, block.firstKeyBytes); cmp == 0 || cmp == 1 {
-			return &block, nil
-		}
+
+func (r *Reader) blockFor(key []byte) (*DataBlock, bool) {
```

Comment:
fwiw, I think dataBlocks already had thread safety issues, since a concurrent request could `.reset()` while another was in progress.